cpp/src/connect/utils.h

```diff
@@ -0,0 +1,240 @@
+class Exception : public std::exception {
+public:
+  /** default ctor */
+  Exception() throw(): std::exception(), msg() {}
+
+  /** copy ctor */
+  Exception(const Exception& src) throw(): std::exception(), msg(src.what()) {
+    collectCallStack();
+  }
+
+  /** ctor from an input message */
+  Exception(const std::string& _msg) throw(): std::exception(), msg(_msg) {
+    collectCallStack();
+  }
+
+  /** dtor */
+  virtual ~Exception() throw() {}
+
+  /** get the message associated with this exception */
+  virtual const char* what() const throw() { return msg.c_str(); }
+
+private:
+  /** message associated with this exception */
+  std::string msg;
+
+  /** append call stack info to this exception's message for ease of debug */
+  // Courtesy: https://www.gnu.org/software/libc/manual/html_node/Backtraces.html
+  void collectCallStack() throw() {
+#ifdef __GNUC__
+    const int MaxStackDepth = 64;
+    void* stack[MaxStackDepth];
+    auto depth = backtrace(stack, MaxStackDepth);
+    std::ostringstream oss;
+    oss << std::endl << "Obtained " << depth << " stack frames" << std::endl;
+    char** strings = backtrace_symbols(stack, depth);
+    if (strings == nullptr) {
+      oss << "But no stack trace could be found!" << std::endl;
+      msg += oss.str();
+      return;
+    }
+    ///@todo: support for demangling of C++ symbol names
+    for (int i = 0; i < depth; ++i) {
+      oss << "#" << i << " in " << strings[i] << std::endl;
+    }
+    free(strings);
+    msg += oss.str();
+#endif // __GNUC__
+  }
+};
+
+/** macro to throw a runtime error */
+#define THROW(fmt, ...)                                                        \
+  do {                                                                         \
+    std::string msg;                                                           \
+    char errMsg[2048];                                                         \
+    std::sprintf(errMsg, "Exception occured! file=%s line=%d: ", __FILE__,     \
+                 __LINE__);                                                    \
+    msg += errMsg;                                                             \
+    std::sprintf(errMsg, fmt, ##__VA_ARGS__);                                  \
+    msg += errMsg;                                                             \
+    throw MLCommon::Exception(msg);                                            \
+  } while (0)
+
+/** macro to check for a conditional and assert on failure */
+#define ASSERT(check, fmt, ...)                                                \
+  do {                                                                         \
+    if (!(check))                                                              \
+      THROW(fmt, ##__VA_ARGS__);                                               \
+  } while (0)
+
+/** check for cuda runtime API errors and assert accordingly */
+#define CUDA_CHECK(call)                                                       \
```

Comment:
@afender 

This may not be an issue for this PR, but we need consistency in error handling. We have utilities/error_utils.h defining CUDA_TRY and this function has CUDA_CHECK, and we brought cuML's error handling mechanism to cuGraph without considering consistency with the remaining part.

I have submitted https://github.com/rapidsai/rmm/issues/83, so we do not need to define our own error handling mechanism for RMM, and it will be better if we can use RAPIDS-wise error handling mechanism instead of creating our own for other parts (and I assume this is a topic for RAPIDS-core, and we need to monitor cuDF to see how they are handling errors).

Reply:
CuGraph error handling needs to improve and bringing cuML's error handling mechanism to cuGraph is increasing this technical debt. I'd expect most libraries to reimplement similar error handling functions. We should :
1. Identify the generic error handling solution for rapids (probably in RAPIDS core).
2. Go through cugraph code and make error handling consistent

I think this will be best addressed as a separate issue/PR.

Reply:
https://github.com/rapidsai/cugraph/issues/231#issuecomment-497463989